### DISCUSSION TITLE
How to use return using the structure logconf.data_received_cb.add_callback(log_stab_callback)?

### DISCUSSION THREAD

**Discussion by Wallat1** [bitcraze/discussions#705] [Q&A]
Hi.

I have the following doubt, how can I use return in the funtion log_stab_callback when is used this structure
logconf.data_received_cb.add_callback(log_stab_callback)?,  the motive of this because I want retun the information of data. The followin  code was taken from 
https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_connect_log_param/

```
import logging
import time

import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie

from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.syncLogger import SyncLogger

# URI to the Crazyflie to connect to
uri = 'radio://0/80/2M/E7E7E7E7E7'

# Only output errors from the logging framework
logging.basicConfig(level=logging.ERROR)

def log_stab_callback(timestamp, data, logconf):
    print('[%d][%s]: %s' % (timestamp, logconf.name, data))

def simple_log_async(scf, logconf):
    cf = scf.cf
    cf.log.add_config(logconf)
    logconf.data_received_cb.add_callback(log_stab_callback)
    logconf.start()
    time.sleep(5)
    logconf.stop()

(...)

if __name__ == '__main__':
    # Initialize the low-level drivers
    cflib.crtp.init_drivers()

    lg_stab = LogConfig(name='Stabilizer', period_in_ms=10)
    lg_stab.add_variable('stabilizer.roll', 'float')
    lg_stab.add_variable('stabilizer.pitch', 'float')
    lg_stab.add_variable('stabilizer.yaw', 'float')

    with SyncCrazyflie(uri, cf=Crazyflie(rw_cache='./cache')) as scf:

        simple_log_async(scf, lg_stab)
```

I'd appreciate your help.

Regards.

**Comment by krichardsson** [comment]
You can find a bunch of examples in https://github.com/bitcraze/crazyflie-lib-python/tree/master/examples

for instance
```python 
def position_callback(timestamp, data, logconf):
    x = data['kalman.stateX']
    y = data['kalman.stateY']
    z = data['kalman.stateZ']
    print('pos: ({}, {}, {})'.format(x, y, z))


def start_position_printing(scf):
    log_conf = LogConfig(name='Position', period_in_ms=500)
    log_conf.add_variable('kalman.stateX', 'float')
    log_conf.add_variable('kalman.stateY', 'float')
    log_conf.add_variable('kalman.stateZ', 'float')

    scf.cf.log.add_config(log_conf)
    log_conf.data_received_cb.add_callback(position_callback)
    log_conf.start()
```

**Reply by Wallat1** [reply]
Hi.

Thank you so much for your help, your example is good, however, if I want to use the variables x, y and z obtained in the function 

```
def position_callback(timestamp, data, logconf):
    x = data['kalman.stateX']
    y = data['kalman.stateY']
    z = data['kalman.stateZ']
    print('pos: ({}, {}, {})'.format(x, y, z))
```
Inside of the function 

```
def start_position_printing(scf):
    log_conf = LogConfig(name='Position', period_in_ms=500)
    log_conf.add_variable('kalman.stateX', 'float')
    log_conf.add_variable('kalman.stateY', 'float')
    log_conf.add_variable('kalman.stateZ', 'float')

    scf.cf.log.add_config(log_conf)
    log_conf.data_received_cb.add_callback(position_callback)
    log_conf.start()
```


or inside of other function to save the values in an append, as I understand in  the function def position_callback(timestamp, data, logconf):  is necessary to include a return or am I wrong?

Regards.

**Reply by krichardsson** [reply]
No, you can not add a return to the callback, it is called from some radio thread somewhere and it will not help you. 

You need to store x, y and z in a way that you can access them from a different thread. One option is global variables and another one is to use a class. Something like 

```python
class MyClass:
    def __init__(self):
        self.x = 0.0
        self.y = 0.0
        self.z = 0.0

    def callback(self, timestamp, data, logconf):
        self.x = data['kalman.stateX']
        # ...

    def other_function(self):
        # Do stuff with x
        print(self.x)
```

I think it might be a good idea to read up on some basic python to understand how the system works and provide the tools for you to solve the problems you are interested in solving.

**Reply by Wallat1** [reply]
Hi.

Thank you very much for the help. I have been using global variables and they are the reason why I decided to ask if it was possible to use return, since with global variables I always get at the beginning the value with which the variable was started and not the current one from the IMU, I am going to test if I am successful with the example provided.

Regards.
